code/modules/station_goals/bluespace_tap.dm

```diff
@@ -0,0 +1,359 @@
+			for(var/i = 1, i <= rand(1, 3), i++)	//freaky shit here, 1-3 freaky portals
+				var/turf/location = locate(x + rand(-5,5), y + rand(-5,5), z)
+				var/mob/living/simple_animal/hostile/spawner/nether/bluespace_tap/portal = new(src)
+				portal.forceMove(location)
+
+
+
+/obj/machinery/power/bluespace_tap/ui_data(mob/user, ui_key = "main", datum/topic_state/state = default_state)
+	var/data[0]
+
+	data["level"] = input_level
+	data["points"] = points
+	data["total_points"] = total_points
+	data["power_use"] = actual_power_usage
+	data["available_power"] = surplus()
+	data["max_level"] = max_level
+	data["emagged"] = emagged
+
+	var/list/listed_items = list()//a list of lists, each inner list equals a datum
+	for(var/key = 1 to length(product_list))
+		var/datum/data/bluespace_tap_product/A = product_list[key]
+		listed_items.Add(list(list(
+				"key" = key,
+				"name" = A.product_name,
+				"price" = A.product_cost)))
+
+
+	data["product"] = listed_items
+	return data
+
+
+/obj/machinery/power/bluespace_tap/attack_hand(mob/user)
+	add_fingerprint(user)
+	ui_interact(user)
+
+
+/obj/machinery/power/bluespace_tap/attack_ai(mob/user)
+	ui_interact(user)
+
+//produces and vends the product with the desired key
+/obj/machinery/power/bluespace_tap/proc/produce(key)
+	var/datum/data/bluespace_tap_product/A = product_list[key]
+	if(!A)	//if called with a bogus key or something, just return
+		message_admins("Yo you fucked up, no product for key [key]")	//debugging, weeeee
+		return
+	if(A.product_cost > points)
+		return
+	points -= A.product_cost
+	A.product_cost *= 1.2
+	playsound(src, 'sound/magic/blink.ogg', 50)
+	new A.product_path(get_turf(src))	//creates product
+
+
+
+//UI stuff below
+
+
+//TODO: Do input/output here
+/obj/machinery/power/bluespace_tap/Topic(href, href_list)
+	if(..())
+		return 1
+	if(href_list["decrease"])
+		decrease_level()
+	else if(href_list["increase"])
+		increase_level()
+	else if(href_list["set"])
+		set_level(input(usr, "Enter new input level (0-[max_level])", "Bluespace Tap Input Control", input_level))
+	else if(href_list["vend"])//it's not really vending as producing, but eh
+		var/key = text2num(href_list["vend"])
+		produce(key)
+
+
+
+/obj/machinery/power/bluespace_tap/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	//stolen from SMES code
+	// update the ui if it exists, returns null if no ui is passed/found
+	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		// the ui does not exist, so we'll create a new() one
+        // for a list of parameters and their descriptions see the code docs in \code\modules\nano\nanoui.dm
+		ui = new(user, src, ui_key, "bluespace_tap.tmpi", "Bluespace Tap", 540, 380)
+		// open the new ui window
+		ui.open()
+		// auto update every Master Controller tick
+		ui.set_auto_update(1)
+
+//emaging allows anyone to use it. Might also add additional options that are unlocked via it.
```

Comment:
not sure it should actually HAVE an access restriction. Not even the BSA has it, need to remove that comment